File: src/rdk_robot_bringup/rdk_robot_bringup/odom_tf_broadcaster.py
import rclpy
from rclpy.node import Node
from nav_msgs.msg import Odometry
from geometry_msgs.msg import TransformStamped
import tf2_ros

class OdomTFBroadcaster(Node):
    def __init__(self):
        super().__init__('odom_tf_broadcaster')
        self.subscription = self.create_subscription(
            Odometry,
            '/odom',
            self.odom_callback,
            10)
        self.tf_broadcaster = tf2_ros.TransformBroadcaster(self)
        self._last_tf = None
        # 以 50Hz 重发最新 TF，防止 ESP32 串口延迟导致 TF 缓冲出现空洞
        self.create_timer(0.02, self._republish_timer)
        self.get_logger().info("Odom TF Broadcaster node has started.")

    def odom_callback(self, msg):
        t = TransformStamped()
        # 使用上位机系统时钟，解决 ESP32 与 RDK X5 时钟不同步问题
        t.header.stamp = self.get_clock().now().to_msg()
        t.header.frame_id = 'odom'
        t.child_frame_id = 'base_footprint'
        t.transform.translation.x = msg.pose.pose.position.x
        t.transform.translation.y = msg.pose.pose.position.y
        t.transform.translation.z = msg.pose.pose.position.z
        t.transform.rotation = msg.pose.pose.orientation
        self._last_tf = t
        self.tf_broadcaster.sendTransform(t)

    def _republish_timer(self):
        # 若已收到过 odom，以当前时间戳重发，保持 TF 缓冲连续
        if self._last_tf is None:
            return
        t = TransformStamped()
        t.header.stamp = self.get_clock().now().to_msg()
        t.header.frame_id = self._last_tf.header.frame_id
        t.child_frame_id = self._last_tf.child_frame_id
        t.transform = self._last_tf.transform
        self.tf_broadcaster.sendTransform(t)

def main(args=None):
    rclpy.init(args=args)
    node = OdomTFBroadcaster()
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    node.destroy_node()
    rclpy.shutdown()

if __name__ == '__main__':
    main()

File: src/rdk_robot_core/src/odom_tf_broadcaster.cpp
#include <chrono>
#include <memory>
#include <mutex>
#include <string>

#include "geometry_msgs/msg/transform_stamped.hpp"
#include "nav_msgs/msg/odometry.hpp"
#include "rclcpp/rclcpp.hpp"
#include "tf2_ros/transform_broadcaster.h"

using namespace std::chrono_literals;

/**
 * @brief OdomTFBroadcaster 节点
 * 订阅 /odom 话题并广播 TF 变换 (odom -> base_footprint)
 * 包含定时重发功能，解决底层硬件(如ESP32)时钟不同步导致的 TF 丢失问题
 */
class OdomTFBroadcaster : public rclcpp::Node {
public:
    OdomTFBroadcaster() : Node("odom_tf_broadcaster") {
        // 初始化 TF 广播器
        tf_broadcaster_ = std::make_unique<tf2_ros::TransformBroadcaster>(*this);

        // 订阅里程计话题
        odom_sub_ = this->create_subscription<nav_msgs::msg::Odometry>(
            "/odom", 10,
            std::bind(&OdomTFBroadcaster::odom_callback, this, std::placeholders::_1));

        // 创建定时器：20Hz (50ms) 重发最新的 TF，确保 TF 缓冲区连续
        timer_ = this->create_wall_timer(
            50ms, std::bind(&OdomTFBroadcaster::republish_timer_callback, this));

        RCLCPP_INFO(this->get_logger(), "Odom TF Broadcaster (C++) started.");
    }

private:
    void odom_callback(const nav_msgs::msg::Odometry::SharedPtr msg) {
        std::lock_guard<std::mutex> lock(mutex_);
        
        auto t = geometry_msgs::msg::TransformStamped();
        t.header.stamp = this->get_clock()->now();
        t.header.frame_id = "odom";
        t.child_frame_id = "base_footprint";

        t.transform.translation.x = msg->pose.pose.position.x;
        t.transform.translation.y = msg->pose.pose.position.y;
        t.transform.translation.z = msg->pose.pose.position.z;
        t.transform.rotation = msg->pose.pose.orientation;

        last_tf_ = t;
        tf_broadcaster_->sendTransform(t);
    }

    void republish_timer_callback() {
        std::lock_guard<std::mutex> lock(mutex_);
        if (last_tf_.header.frame_id.empty()) {
            return;
        }

        // 更新时间戳并重发，解决某些情况下 TF 缓冲出现的空洞
        last_tf_.header.stamp = this->get_clock()->now();
        tf_broadcaster_->sendTransform(last_tf_);
    }

    // ROS 2 相关成员
    rclcpp::Subscription<nav_msgs::msg::Odometry>::SharedPtr odom_sub_;
    std::unique_ptr<tf2_ros::TransformBroadcaster> tf_broadcaster_;
    rclcpp::TimerBase::SharedPtr timer_;

    // 数据缓存与线程安全
    geometry_msgs::msg::TransformStamped last_tf_;
    std::mutex mutex_;
};

int main(int argc, char* argv[]) {
    rclcpp::init(argc, argv);
    rclcpp::spin(std::make_shared<OdomTFBroadcaster>());
    rclcpp::shutdown();
    return 0;
}